File: ios/renderer/ListRenderer.m
#import "ListRenderer.h"
#import "BlockquoteRenderer.h"
#import "MarkdownASTNode.h"
#import "ParagraphStyleUtils.h"
#import "RenderContext.h"
#import "RendererFactory.h"
#import "StyleConfig.h"

@implementation ListRenderer {
  __weak RendererFactory *_rendererFactory;
  StyleConfig *_config;
  BOOL _isOrdered;
}

- (instancetype)initWithRendererFactory:(RendererFactory *)factory config:(StyleConfig *)config isOrdered:(BOOL)ordered
{
  if (self = [super init]) {
    _rendererFactory = factory;
    _config = config;
    _isOrdered = ordered;
  }
  return self;
}

- (void)renderNode:(MarkdownASTNode *)node into:(NSMutableAttributedString *)output context:(RenderContext *)context
{
  if (!context)
    return;

  const NSInteger prevDepth = context.listDepth;
  const ListType prevType = context.listType;
  const NSInteger prevNum = context.listItemNumber;

  const NSUInteger startLocation = output.length;
  NSUInteger contentStart = startLocation;

  if (prevDepth == 0) {
    // Apply top margin for root-level list
    contentStart += applyBlockSpacingBefore(output, startLocation, _config.listStyleMarginTop);
  } else if (output.length > 0 && ![output.string hasSuffix:@"\n"]) {
    // Ensure nested lists start on a new line
    [output appendAttributedString:kNewlineAttributedString];
  }

  context.listDepth = prevDepth + 1;
  context.listType = _isOrdered ? ListTypeOrdered : ListTypeUnordered;
  context.listItemNumber = 0; // Reset counter for this specific list level

  [context setBlockStyle:_isOrdered ? BlockTypeOrderedList : BlockTypeUnorderedList
                    font:_config.listStyleFont
                   color:_config.listStyleColor
            headingLevel:0];

  @try {
    [_rendererFactory renderChildrenOfNode:node into:output context:context];
  } @finally {
    context.listDepth = prevDepth;
    context.listType = prevType;
    context.listItemNumber = prevNum;

    if (prevDepth == 0) {
      [context clearBlockStyle];

      // Apply bottom margin for root-level list
      applyBlockSpacingAfter(output, _config.listStyleMarginBottom);
    }
  }
}

@end

File: packages/ios-enriched-markdown/Sources/EnrichedMarkdown/Rendering/Renderers/ListRenderer.swift
import UIKit

final class ListRenderer: NodeRenderer {
    private let factory: RendererFactory
    private let config: MarkdownStyleConfig
    private let isOrdered: Bool

    init(factory: RendererFactory, config: MarkdownStyleConfig, isOrdered: Bool) {
        self.factory = factory
        self.config = config
        self.isOrdered = isOrdered
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        let prevDepth = context.listDepth
        let prevType = context.listType
        let prevNumber = context.listItemNumber
        let startLocation = output.length

        if prevDepth == 0 {
            ParagraphStyleHelpers.ensureStartingOnNewLine(in: output)
            _ = ParagraphStyleHelpers.applyBlockSpacingBefore(
                to: output,
                at: startLocation,
                marginTop: config.list.marginTop ?? 0
            )
        } else if output.length > 0, !output.string.hasSuffix("\n") {
            output.append(ParagraphStyleHelpers.newline)
        }

        context.listDepth = prevDepth + 1
        context.listType = isOrdered ? .ordered : .unordered
        context.listItemNumber = 0

        let font = config.list.font ?? UIFont.preferredFont(forTextStyle: .body)
        let color = config.list.foregroundColor ?? UIColor.label
        context.setBlockStyle(
            font: font,
            color: color,
            blockType: isOrdered ? .orderedList : .unorderedList
        )

        factory.renderChildren(of: node, into: output, context: context)

        context.listDepth = prevDepth
        context.listType = prevType
        context.listItemNumber = prevNumber

        if prevDepth == 0 {
            context.clearBlockStyle()
            if let marginBottom = config.list.marginBottom, marginBottom > 0 {
                ParagraphStyleHelpers.applyBlockSpacingAfter(to: output, marginBottom: marginBottom)
            }
        }
    }
}
